### PR TITLE
[netcore] Implement missing Bmi1/Bmi2 intrinsics

### PR DESCRIPTION
**Pull request by EgorBo**
Implement all missing Bmi1/Bmi2 intrinsics except `MultiplyNoFlags` (no idea how to properly implement it yet).

Tested on a CPU with both bmi1 and bmi2 sets:
```csharp
using System;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics.X86;

public class Program
{
    public static void Main(string[] args)
    {
        Console.WriteLine(Bmi1.IsSupported);
        Console.WriteLine(Bmi2.IsSupported);
        Console.WriteLine();

        // BMI1
        Console.WriteLine(AndNot1(1234, 3));
        Console.WriteLine(AndNot2(1234, 3));
        Console.WriteLine(BitFieldExtract3(1234, 3));
        Console.WriteLine(BitFieldExtract4(1234, 3));
        Console.WriteLine(ExtractLowestSetBit7(1234));
        Console.WriteLine(ExtractLowestSetBit8(1234));
        Console.WriteLine(GetMaskUpToLowestSetBit9(1234));
        Console.WriteLine(GetMaskUpToLowestSetBit10(1234));
        Console.WriteLine(ResetLowestSetBit11(1234));
        Console.WriteLine(ResetLowestSetBit12(1234));
        Console.WriteLine(TrailingZeroCount13(1234));
        Console.WriteLine(TrailingZeroCount14(1234));

        // BMI2
        Console.WriteLine(ZeroHighBits15(1234, 3));
        Console.WriteLine(ZeroHighBits16(1234, 3));
        Console.WriteLine(ParallelBitDeposit17(1234, 3));
        Console.WriteLine(ParallelBitDeposit18(1234, 3));
        Console.WriteLine(ParallelBitExtract19(1234, 3));
        Console.WriteLine(ParallelBitExtract20(1234, 3));
    }

    // BMI1
    [MethodImpl(MethodImplOptions.NoInlining)] static uint AndNot1(uint a, uint c) => Bmi1.AndNot(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong AndNot2(ulong a, ulong c) => Bmi1.X64.AndNot(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint BitFieldExtract3(uint a, ushort c) => Bmi1.BitFieldExtract(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong BitFieldExtract4(ulong a, ushort c) => Bmi1.X64.BitFieldExtract(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint ExtractLowestSetBit7(uint a) => Bmi1.ExtractLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong ExtractLowestSetBit8(ulong a) => Bmi1.X64.ExtractLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint GetMaskUpToLowestSetBit9(uint a) => Bmi1.GetMaskUpToLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong GetMaskUpToLowestSetBit10(ulong a) => Bmi1.X64.GetMaskUpToLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint ResetLowestSetBit11(uint a) => Bmi1.ResetLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong ResetLowestSetBit12(ulong a) => Bmi1.X64.ResetLowestSetBit(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint TrailingZeroCount13(uint a) => Bmi1.TrailingZeroCount(a);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong TrailingZeroCount14(ulong a) => Bmi1.X64.TrailingZeroCount(a);

    // BMI2
    [MethodImpl(MethodImplOptions.NoInlining)] static uint ZeroHighBits15(uint a, uint c) => Bmi2.ZeroHighBits(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong ZeroHighBits16(ulong a, ulong c) => Bmi2.X64.ZeroHighBits(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint ParallelBitDeposit17(uint a, uint c) => Bmi2.ParallelBitDeposit(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong ParallelBitDeposit18(ulong a, ulong c) => Bmi2.X64.ParallelBitDeposit(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static uint ParallelBitExtract19(uint a, uint c) => Bmi2.ParallelBitExtract(a, c);
    [MethodImpl(MethodImplOptions.NoInlining)] static ulong ParallelBitExtract20(ulong a, ulong c) => Bmi2.X64.ParallelBitExtract(a, c);
}
```

Codegen matches CoreCLR

### CODE REVIEW

File: mono/mini/simd-intrinsics-netcore.c [hunk 1]
```diff
@@ -686,9 +706,6 @@ emit_x86_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature
 		}
 	}
 	if (!strcmp (class_name, "Bmi2") || (!strcmp (class_name, "X64") && cmethod->klass->nested_in && !strcmp (m_class_get_name (cmethod->klass->nested_in), "Bmi2"))) {
-		// We only support the subset used by corelib
-		if (m_class_get_image (cfg->method->klass) != mono_get_corlib ())
-			return NULL;
```
Comment (filipnavara) [inline mono/mini/simd-intrinsics-netcore.c]: I don't think you can safely remove this until `SN_MultiplyNoFlags` is implemented. `IsSupported` will return `true` but the actual method will end up with throwing `PlatformNotSupportedException` (or similar).

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: yeah... but it unlocks Bmi2 for users, MultiplyNoFlags is currently problematic in CoreCLR: https://github.com/dotnet/coreclr/issues/21899

Comment (filipnavara) [reply mono/mini/simd-intrinsics-netcore.c]: I do realize that but even an unoptimized fallback would be better option. The linked issue shows that there are actually people trying to use it. For those people the situation with `Bmi2.IsSupported == true` would suddenly start throwing exceptions in perfectly legitimate code.

Comment (vargaz) [reply mono/mini/simd-intrinsics-netcore.c]: Yes, that check should only be removed if all the methods are supported.

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: @filipnavara @vargaz ok

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/simd-intrinsics-netcore.c [hunk 2]
```diff
@@ -721,14 +721,14 @@ emit_x86_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature
...
-		case SN_ZeroHighBits:
+		//case SN_MultiplyNoFlags:
+			//// TODO: implement using _mulx_u32/u64:
+			//// ulong MultiplyNoFlags(ulong left, ulong right)
+			//// ulong MultiplyNoFlags(ulong left, ulong right, ulong* low) => MultiplyNoFlags(left, right, low);
+			//// uint MultiplyNoFlags(uint left, uint right)
+			//// uint MultiplyNoFlags(uint left, uint right, uint* low)
+			//return NULL;
+		//case SN_ZeroHighBits:
```
Comment (filipnavara) [inline mono/mini/simd-intrinsics-netcore.c]: One line too much of comment?

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: you mean `////` - it's for when I(someone) will uncomment the whole block - the comments will remain comments :)

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: had to comment SN_MultiplyNoFlags for LLVM AOT mode (it crashed with vred != -1 assert)

Comment (filipnavara) [reply mono/mini/simd-intrinsics-netcore.c]: You commented out `case SN_ZeroHighBits` too.

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: oops 🙂

**Comment by imhameed** [comment]
> 
> 
> Implement all missing Bmi1/Bmi2 intrinsics except `MultiplyNoFlags` (no idea how to properly implement it yet).

There are two fallbacks that look sensible to me:

1. Emit LLVM IR that may (or may not) generate `mulx`, like happens when using `_mulx_u64` or `_mulx_u32` with Clang:
https://github.com/llvm/llvm-project/blob/release/6.x/clang/lib/Headers/bmi2intrin.h#L72

2. Emit IR that contains an inline assembly blob; see `do_mulx_64_iasm`:
https://godbolt.org/z/bgJss9

**Pull request merged**